micro-benchmarks/DRB144-gpu-atomiccritical-orig-no.c

```diff
@@ -0,0 +1,34 @@
+/*
+!!!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~!!!
+!!! Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
+!!! and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
+!!!
+!!! SPDX-License-Identifier: (BSD-3-Clause)
+!!!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~!!!
+ */
+
+/*
+The condition check at line 26:8 is critical and increment at line number 28 is atomic for the variable
+var@28:7. Therefore, there are no data race pairs.
+*/
+
+#include <stdio.h>
+#include <omp.h>
+#define N 100
+
+int var = 0;
+
+int main(){
+  #pragma omp target map(tofrom:var) device(0)
+  #pragma omp teams distribute parallel for
+  for (int i=0; i<N*2; i++){
+    #pragma omp critical
```

Comment:
I agree that this needs to be investigated. Synchronization is indeed only guaranteed within a same team. @xintin

Reply:
@chunhualiao, @jprotze ,

To understand it better, 
1. it will encounter teams - now say we have a league of n teams created. 
2. enter the for loop
3. For any 10 (if N==10) threads, this for loop will execute, and one thread from each team will update the var.
4. The critical construct will restrict to a single thread at a time.
5. The atomic construct ensures that a specific storage location is accessed atomically, rather than exposing it to the possibility of multiple, simultaneous reading and writing threads that may result in indeterminate values. 

So, in my understanding, though critical may not synchronize across teams, critical+atomic will avoid any data race in var. 

Kindly feel free to correct me here.

Reply:
The critical protects the var in the if condition to be changed by another thread in the same team. 
The read of var still is a data race with the increment in another team. Therefore, this code is another example for an inter-team data race.

In an implementation, where teams are processed in a pipeline one by one, one will not be able to observe the race.
When you think of a multi-socket system with multiple NUMA domains, each NUMA domain executes one team. To avoid the cost of inter-NUMA communication, the implementation will restrict the effect of locks to the team. In this case, other teams will still be able to increment var, while one thread reads var.

Reply:
Oh, I see. Makes sense.

 ```
 #pragma omp target map(tofrom:var) device(0)
  #pragma omp distribute parallel for
  for (int i=0; i<N*2; i++){
    #pragma omp critical
    if(var<N){
      #pragma omp atomic
      var++;
    }
  }
``` 

Do you suggest this should be fine? considering no team at all.

Reply:
Your proposed code has no data race.

As an alternative, the following reduction should also resolve the data race:
```C
  #pragma omp target map(tofrom:var) device(0)
  #pragma omp teams reduction(+:var)
  #pragma omp distribute parallel for
  for (int i=0; i<N*2; i++){
    #pragma omp critical
    if(var<N){
      #pragma omp atomic
      var++;
    }
  }
```

Reply:
Correct. Since these are -no kernels. I will use both of these alternatives. Thanks.

Reply:
@jprotze Probably, you would like to have a look here. The reduction clause you have proposed, I tried that, and I am getting 20 as the output. I do not think that is correct. It should be 10. I mean, it may/may not have data race but is it the correct execution?

Reply:
Since each team works on a private instance of var, the result will be N * number of teams, or number of total loop iterations whatever is smaller.

Reply:
Agreed, but will that be an accepted kernel here? Would you recommend reduction(+: var) or num_teams(1), considering it is a -no scenario? In both, there won't be any data race.

Reply:
@chunhualiao @jprotze 

Please confirm the solution to be used for this -no kernel. 
1. num_teams(1) -- execution on the host is giving output the same as on the device. 
2. reduction(+: var) -- execution result will differ on the host and on-device due to teams.